core/src/main/java/fr/sncf/osrd/envelope_sim/TrainPhysicsIntegrator.java

```diff
@@ -0,0 +1,155 @@
+    ) {
+        var integrator = new TrainPhysicsIntegrator(rollingStock, path, action, directionSign);
+        var halfStep = timeStep / 2;
+        var step1 = integrator.step(halfStep, initialLocation, initialSpeed);
+        var step2 = integrator.step(halfStep, initialLocation + step1.positionDelta, step1.speed);
+        var step3 = integrator.step(timeStep, initialLocation + step2.positionDelta, step2.speed);
+        var step4 = integrator.step(timeStep, initialLocation + step3.positionDelta, step3.speed);
+
+        var meanAcceleration = (
+                step1.acceleration
+                + 2 * step2.acceleration
+                + 2 * step3.acceleration
+                + step4.acceleration
+        ) / 6.;
+        return newtonStep(timeStep, initialSpeed, meanAcceleration, directionSign);
+    }
+
+    private IntegrationStep step(double timeStep, double location, double speed) {
+
+        double tractionForce = 0;
+        double brakingForce = 0;
+        double maxTractionForce = rollingStock.getMaxEffort(speed);
+        double rollingResistance = rollingStock.getRollingResistance(speed);
+        double weightForce = getWeightForce(rollingStock, path, location);
+
+        if (action == Action.ACCELERATE)
+            tractionForce = maxTractionForce;
+
+        if (action == Action.BRAKE)
+            brakingForce = rollingStock.getMaxBrakingForce(speed);
+
+        if (action == Action.MAINTAIN) {
+            if (canMaintainSpeed(rollingResistance, weightForce, speed, maxTractionForce))
+                return newtonStep(timeStep, speed, 0.0, directionSign);
+            tractionForce = maxTractionForce;
+        }
+
+        double acceleration = computeAcceleration(rollingStock, rollingResistance,
+                weightForce, speed, tractionForce, brakingForce, directionSign);
+        return newtonStep(timeStep, speed, acceleration, directionSign);
+    }
+
+    private boolean canMaintainSpeed(double rollingResistance,
+                                     double weightForce,
+                                     double speed,
+                                     double maxTractionForce) {
+        var testAcceleration = computeAcceleration(rollingStock, rollingResistance,
+                weightForce, speed, maxTractionForce, 0, directionSign);
+        return testAcceleration >= 0;
+    }
+
+    private static double getWeightForce(PhysicsRollingStock rollingStock, PhysicsPath path, double headLocation) {
+        var tailLocation = Math.max(0, headLocation - rollingStock.getLength());
+        var averageSlope = path.getAverageSlope(tailLocation, headLocation);
+        // get an angle from a meter per km elevation difference
+        // the curve's radius is taken into account in meanTrainGrade
+        var angle = Math.atan(averageSlope / 1000.0);  // from m/km to m/m
+        return -rollingStock.getMass() * 9.81 * Math.sin(angle);
+    }
+
+    private static double computeAcceleration(
```

Comment:
Lots of arguments here. It would make a lot of sense to have only forces and directionsign as arguments of this function